manticore/core/executor.py

```diff
@@ -69,184 +46,276 @@ def newFunction(self, *args, **kw):
+
+class Random(Policy):
+    def __init__(self):
+        super(Random, self).__init__()
+
+    def features(self, state):
+        ''' Save state features for prioritization before a state is stored '''
+        pass
+
+    def priority(self, state_id):
+        ''' A numeric value representing likelihood to reach the interesting program spot '''
+        return 1.0
 
-        self.loading_time = 0
-        self.saving_time = 0
-        self.solver_time = 0
-        for (func_file, _, func_name), (_, _, _, func_time, _) in raw_stats.stats.iteritems():
-            if func_name == '_getState':
-                self.loading_time += func_time
-            elif func_name == '_putState':
-                self.saving_time += func_time
-            elif func_file.endswith('solver.py') and 'setstate' not in func_name and 'getstate' not in func_name and 'ckl' not in func_name:
-                self.solver_time += func_time
 
 class Executor(object):
     '''
-    The executor guides the execution of a single state, handles state forking 
-    and selection, maintains run statistics and handles all exceptional
-    conditions (system calls, memory faults, concretization, etc.)
+    The executor guides the execution of an initial state or a paused previous run. 
+    It handles all exceptional conditions (system calls, memory faults, concretization, etc.)
     '''
 
-    def shutdown(self):
-        with self._lock:
-            self._shutdown.value+=1
-            ret = self._shutdown.value
-            self._lock.notify_all()
-        return ret
-
-    def isShutdown(self):
-        return self._shutdown.value != 0
-
-    def __init__(self, initial, workspace=None, policy='random', **options):
-        assert os.path.isdir(workspace)
+    def __init__(self, initial=None, workspace=None, policy='random', **options):
+        assert os.path.isdir(workspace), 'Workspace must be a directory'
         self.workspace = workspace
         logger.debug("Workspace set: %s", self.workspace)
 
-        self.max_states = options.get('maxstates', 0)
-        self.max_storage = options.get('maxstorage', 0)
-        self._dump_every = options.get('dumpafter', 0)
-        self._profile = cProfile.Profile()
-        self.profiling = options.get('dumpstats', False)
-
-        # Signals
-        self.will_execute_pc = Signal()
-        self.will_fork = Signal()
-
-        self.policy_order = 'max'
-        if policy[0] in '+-':
-            self.policy_order = {'+':'max', '-':'min'}[policy[0]]
-            policy = policy[1:]
-        assert policy in ['random','adhoc','uncovered','dicount','icount','syscount','depth','bucket']
-        self.policy = policy
-
-        self._states = manager.dict()
-        self._test_number = manager.Value('i', 0 )
+        # Signals / Callbacks handlers will be invoked potentially at differeent 
+        # worker processes. State provides a local context to save data.
+
+        #Executor
+        self.will_start_run = Signal()
```

Comment:
are we going a little signal crazy here? a lot of these seem unused. again, i think we should have documented use cases first